### ISSUE TITLE
support Json Null value and Optionality

### ISSUE THREAD

**Issue by ekeren** [winglang/wing#1819]
### Community Note

> Please vote by adding a 👍 reaction to the issue to help us prioritize.
> If you are interested to work on this issue, please leave a comment.

### Feature Spec

According to https://www.json.org/json-en.html Json also supports Json Null value
This means that `{"k1": null}` is a valid json that is different from `{}` 🤯 

Currently (0.5.129) we do not support this in Wing

**Note**:  I assume that this feature will not be used a lot, but it is important for completeness   

### Use Cases

We want to consider supporting the following capabilities, Here is an initial proposal:
###  Allow developer to define a Json Null Value
**proposal:**  `Json.Null` (not sure about the capital N)
```ts (wing)
let j = Json { 
  k1: Json.Null
}'
```
### support the `?` operator 
The `?` operator should work for null value and for no value (with no distinction)
```ts (wing)
let j = Json { 
  k1: Json.Null
  k2: "something"
}'
assert(j.k0? == false);
assert(j.k1? == false);
assert(j.k2? == true);
```
### Support `has_key` for distinguishing between Json.Null and no Value
We want to allow Wing distinguish between a null value for key `k1` vs no `k1` using `has_key` static `Json` function

```ts (wing)
let j = Json { 
  k1: Json.Null
  k2: "something"
}'
assert(Json.has_key("k0") == false);
assert(Json.has_key("k1") == true);
assert(Json.has_key("k0") == true);
```

### Component

Language Design

**Comment by yoav-steinberg** [comment]
I really like the static `Json.Null` which hides the internal type system's implementation of `null` for json.

I don't see why we need the `?` operator and it also seems to mix up with wing's optionality syntax. I think this should be enough:
```wing
let j = Json { 
  k1: Json.Null
  k2: "something"
}'
assert(j.has_key("k0") == false);
assert(j.has_key("k1") == true);
assert(j.k1 == Json.Null);
```

**Comment by ekeren** [comment]
>  I don't see why we need the ? operator and it also seems to mix up with wing's optionality synta

My intent was actually to extend wing's `?` operator to treat `Json.Null` as `null` 

I see the disadvantage of treating the `Json.Null` special for`?` operator, but from pragmatic standpoint I think we need a better solution then what you are suggesting 

Think about this code: 
```ts (wing) 
// some_bad_json_producer.com might return `{name: null}`or `{}` or `{name: "John"}`
let j = read_json("https://some_bad_json_producer.com").mut_copy(); 

// I need to always check nullability of an object and key existence 
if j1.name? && j1.name != Json.Null {
  j1.name = 'default name';
}
```

I think that it is **not uncommon** to get this `{k: null}` response from services, keep in mind that in Javascript these are almost the same, so test don't usually fail... (I think I've seen this kind of result coming from mongo) 

For me, it boils down to the question:
**How much is this common `{k: null}`**?

**Comment by yoav-steinberg** [comment]
It can even be worse, the evil producer might return a `{name: 6}` or `{name: {first: "Clark", last: "Kent" }}`. 
My opinion is that coping with these things should be application specific. The best tools are probably json path or jq and json schema validation which I'd love to integrate into the language. But we should also provide primitives as part of our Json class API. I wouldn't go into fancy syntax specifically for Json at this point.

**Comment by Chriscbr** [comment]
@hasanaburayyan I don't think we can have methods directly on Json objects since there could be a key on the object named `try_get`, right? It might have to be like:

```js
if Json.try_get(j1, "name") {
	j1.name = 'default name';
}
```

**Comment by ekeren** [comment]
I had an issue on my example,  it should be 
```ts
if !j1.name? || j1.name == Json.Null {
  j1.name = 'default name';
}
```

and I think that @hasanaburayyan example should also be prefix with `!` and suffixed with `?`, hence:
```ts
if !j1.try_get("name")? {
	j1.name = 'default name';
}
```

**Comment by ekeren** [comment]
added @hasanaburayyan's option to the description

**Comment by ekeren** [comment]
@eladb eventually we need your decision here

**Comment by eladb** [comment]
Seems like a great direction. I'll dig into the details next week ❤️

**Comment by eladb** [comment]
@staycoolcall911 I think we should bump this to P1. Without this it's almost impossible to work with JSON...
